### ISSUE TITLE
Offsets here vs pytorch

### ISSUE THREAD

**Issue by amitport** [yahoo/weighted_fields_recsys#1]
Hi,

Thanks for publishing! This looks really useful.

I noticed that offsets here (which are offsets to a bag ending position) are defined differently than pytorch's embedding bags offsets (which are offsets to a bag starting position, except the optional last offset).

Would you consider aligning the definition with the pytorch one, which seems to make more sense as "offset"?

Thank you!

**Comment by alexshtf** [comment]
Hi Amit.

This is on purpose - by specifying where each bag _ends_ rather than _begins_, we can support a different number of embeddings per sample in a mini-batch, since we can ignore the padding tail of each sample. If you have an alternative design in mind, which allows using a different number of embeddings per training sample, please write it here.

**Comment by amitport** [comment]
It's the same info only encoded differently, no? It's just that begining of a bag correspond more the the term "offset" IMO as it is used in pytorch.
```
pytorch_offsets = [0] + (offsets[:-1] + 1)
```

**Comment by alexshtf** [comment]
How, in PyTorch, do you specify where the last bag ends? For example, suppose we have two bags in the following **padded** sample (padded, because it's a part of a mini-batch, where each sample has a different number of embeddings):
```
[1, 5, 8, 10, 2, 0, 0, 0]
```
I want the first two embeddings [1, 5] to belong to one bag, and the next three  [8, 10, 2] to belong to the another bag, and ignore the padding tail. In our design we can specify [1, 4] as the bag endpoints. How would you do it with the convention used by torch.nn.EmbeddingBag, where you specify where each bag begins? How would you make it ignore the tail?

**Comment by amitport** [comment]
I don't think pytorch supports this without changing the input a bit. In any case, pytorch offsets are offsets and here they are inclusive endpoints. Seems confusing, but this is just a suggestion. Thank you for publishing.

**Comment by alexshtf** [comment]
Yes, I understand.  It was easier to do experiments with this design we chose, since we could easily construct the input by concatenating field embedding indices, and then constructing a mini-batch in a custom collate_fn using using `torch.nn.utils.rnn.pad_sequence`, like this:
```
    indices = ... # a list of int tensors
    weights = ... # a list of float tensors
    indices = torch.nn.utils.rnn.pad_sequence(indices, batch_first=True, padding_value=0)
    weights = torch.nn.utils.rnn.pad_sequence(weights, batch_first=True, padding_value=0.)
    ...
```

The idea you suggested, using the convention chosen by `torch.nn.EmbeddingBag` is also possible, since we can just use zero weights for the embeddings we don't need, and that way support a variable number of embeddings in each bag. However, it makes mini-batch construction in your DataLoader a bit harder, and also wastes time on multiplying embeddings you won't use by a zero weight, which prolongs the time we invest in running experiments by a bit.

I think we can solve it by renaming the `offsets` keyword to `endpoints`, and using another `offsets` keyword for the convention used by torch.nn.EmbeddingBag. That way both conventions will be supported, and it will make adoption by people who want the PyTorch convention easier, but also our experimentation easier. Of course, specifying both `offsets` and `endpoints` will be considered an error.

**Comment by amitport** [comment]
That sound great thank you!
